web/autograder_service.go

```diff
@@ -605,37 +605,39 @@ func (s *AutograderService) UpdateSubmission(ctx context.Context, in *pb.UpdateS
 	return &pb.Void{}, err
 }
 
-// RebuildSubmission rebuilds the submission with the given ID
-// Access policy: any user.
-func (s *AutograderService) RebuildSubmission(ctx context.Context, in *pb.RebuildRequest) (*pb.Submission, error) {
-	if !s.isValidSubmission(in.GetSubmissionID()) {
-		s.logger.Errorf("ApproveSubmission failed: submitter has no access to the course")
-		return nil, status.Error(codes.PermissionDenied, "submitter has no course access")
-	}
-	submission, err := s.rebuildSubmission(in)
-	if err != nil {
-		s.logger.Errorf("RebuildSubmission failed: %v", err)
-		return nil, status.Error(codes.InvalidArgument, "failed to rebuild submission")
-	}
-	return submission, nil
-}
-
-// RebuildSubmissions runs tests for all submissions for the given assignment ID.
+// RebuildSubmissions runs tests for a single submission if the request contain submission ID
+// or for all submissions for the given assignment if the request contains course ID.
 // Access policy: Teacher of CourseID.
-func (s *AutograderService) RebuildSubmissions(ctx context.Context, in *pb.AssignmentRequest) (*pb.Void, error) {
-	usr, err := s.getCurrentUser(ctx)
-	if err != nil {
-		s.logger.Errorf("RebuildSubmissions failed: authentication error: %v", err)
-		return nil, ErrInvalidUserInfo
-	}
-	if !s.isTeacher(usr.ID, in.GetCourseID()) {
-		s.logger.Error("RebuildSubmissions failed: user is not teacher")
-		return nil, status.Error(codes.PermissionDenied, "only teachers can rebuild all submissions")
-	}
-	if err := s.rebuildSubmissions(in); err != nil {
-		s.logger.Errorf("RebuildSubmissions failed: %v", err)
-		return nil, status.Error(codes.InvalidArgument, "failed to rebuild submissions")
+// TODO: Access policy: any for SubmissionID type ?
```

Comment:
How about this?:
```go
// Access policy: Teacher of CourseID or student / group owner of SubmissionID.
```

Reply:
Revising my own comment. Based on what we wrote for `GetSubmissions`, we should say something like this:
```go
// Access policy:
// Current User if Owner of submission,
// Current User if member of group for group submission,
// Teacher of CourseID.
```

Reply:
Note that neither the old or redesigned frontend has implemented student access for rebuilding submissions.

Currently the `s.isValidSubmission` checks:
- If the submission belongs to an enrollment with status `Student` or higher
- If the submission belongs to a group that belongs to the course.

However, the submission checked is retrieved from the database based on the received submission ID in the received request.

With this in mind, a student could in theory trigger a gRPC call to `RebuildSubmissions` with a request containing an arbitrary submission ID. As long as the conditions listen above are fulfilled, this could trigger a rebuild of a submission not belonging to the student.

Either we:
- Allow students to rebuild their own submissions, requiring minor modifications to `isValidSubmission`
    - Minor modifications to the frontend is required as well
- Disallow students from rebuilding their own submissions. Done by simply moving the `isTeacher` check to the top.

Reply:
Great. You are paying attention. We should disallow students to rebuild since this feature was available before and was misused when or our tests had bugs etc.